src/Common/remapExecutable.cpp

```diff
@@ -0,0 +1,211 @@
+#if defined(__linux__) && defined(__amd64__) && defined(__SSE2__) && !defined(SANITIZER) && defined(NDEBUG)
+
+#include <sys/mman.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+
+#include <emmintrin.h>
+
+#include <utility>
+
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/hex.h>
+#include <Common/Exception.h>
+#include <IO/ReadBufferFromFile.h>
+#include <IO/ReadHelpers.h>
+
+#include "remapExecutable.h"
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int CANNOT_ALLOCATE_MEMORY;
+}
+
+
+namespace
+{
+
+uintptr_t readAddressHex(DB::ReadBuffer & in)
+{
+    uintptr_t res = 0;
+    while (!in.eof())
+    {
+        if (isHexDigit(*in.position()))
+        {
+            res *= 16;
+            res += unhex(*in.position());
+            ++in.position();
+        }
+        else
+            break;
+    }
+    return res;
+}
+
+
+/** Find the address and size of the mapped memory region pointed by ptr.
+  */
+std::pair<void *, size_t> getMappedArea(void * ptr)
+{
+    using namespace DB;
+
+    uintptr_t uintptr = reinterpret_cast<uintptr_t>(ptr);
+    ReadBufferFromFile in("/proc/self/maps");
+
+    while (!in.eof())
+    {
+        uintptr_t begin = readAddressHex(in);
+        assertChar('-', in);
+        uintptr_t end = readAddressHex(in);
+        skipToNextLineOrEOF(in);
+
+        if (begin <= uintptr && uintptr < end)
+            return {reinterpret_cast<void *>(begin), end - begin};
+    }
+
+    throw Exception("Cannot find mapped area for pointer", ErrorCodes::LOGICAL_ERROR);
+}
+
+
+__attribute__((__noinline__)) int64_t our_syscall(...)
+{
+     __asm__ __volatile__ (R"(
+        movq %%rdi,%%rax;
+        movq %%rsi,%%rdi;
+        movq %%rdx,%%rsi;
+        movq %%rcx,%%rdx;
+        movq %%r8,%%r10;
+        movq %%r9,%%r8;
+        movq 8(%%rsp),%%r9;
+        syscall;
+        ret
+    )" : : : "memory");
+    return 0;
+}
+
+
+__attribute__((__noinline__)) void remapToHugeStep3(void * scratch, size_t size, size_t offset)
+{
+    /// The function should not use the stack, otherwise various optimizations, including "omit-frame-pointer" may break the code.
```

Comment:
Just my curious. Is that why we can't clean up after jump? It's like https://github.com/ClickHouse/ClickHouse/pull/14685/files#diff-aec223e38b51ea8306ee7a57c508315dR188.

Reply:
Sorry, did not get the question completely... how to clean up after jump?

Reply:
Probably we also rely on the fact that the function does not have prologue.
(It also does not permit debug build.)
But we also can compile this function without optimiations (using some attribute).